Lib/http/server.py

```diff
@@ -868,21 +894,37 @@ def translate_path(self, path):
             path += '/'
         return path
 
-    def copyfile(self, source, outputfile):
-        """Copy all data between two file objects.
+    def copyfile(self, source, outputfile, *, range=None):
+        """Copy all data between two file objects if range is None.
+        Otherwise, copy data between two file objects based on the
+        inclusive range (start, end).
 
         The SOURCE argument is a file object open for reading
-        (or anything with a read() method) and the DESTINATION
-        argument is a file object open for writing (or
-        anything with a write() method).
+        (or anything with read() and seek() method) and the
+        DESTINATION argument is a file object open for writing
+        (or anything with a write() method).
 
         The only reason for overriding this would be to change
         the block size or perhaps to replace newlines by CRLF
         -- note however that this the default server uses this
         to copy binary data as well.
 
         """
-        shutil.copyfileobj(source, outputfile)
+        if range is None:
+            shutil.copyfileobj(source, outputfile)
+        else:
+            start, end = range
+            length = end - start + 1
+            source.seek(start)
+            while True:
+                if length <= 0:
+                    break
+                buf = source.read(min(length, shutil.COPY_BUFSIZE))
+                if not buf:
+                    break
```

Comment:
```suggestion
                    raise EOFError('File shrank after size was checked')
```

I believe this should make the server kill the HTTP connection, indicating the response failed halfway and the connection is no longer usable. Otherwise, if the connection is persistent (protocol = 'HTTP/1.1'), the server will wait for another request while the client is still expecting the rest of the promised Content-Length response bytes.

This would be a bug with the existing non-range code path as well.